### PR TITLE
8266854: LibraryCallKit::inline_preconditions_checkIndex modifies control flow even if the intrinsic bailed out

### PR DESCRIPTION
**Pull request by sviswa7**
LibraryCallKit::inline_preconditions_checkIndex can result in the following assert sometimes:
"# assert(ctrl == kit.control()) failed: Control flow was added although the intrinsic bailed out"

Consider the following code snippet:
...
set_control(_gvn.transform(new IfTrueNode(rc)));
{
PreserveJVMState pjvms(this);
set_control(_gvn.transform(new IfFalseNode(rc)));
uncommon_trap(Deoptimization::Reason_range_check,
Deoptimization::Action_make_not_entrant);
}
..
Here the control is being modified by set_control and afterwards a bailout happens.

The fix is as per discussion on https://github.com/openjdk/jdk/pull/3958.

Best Regards,
Sandhya

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266854](https://bugs.openjdk.java.net/browse/JDK-8266854): LibraryCallKit::inline_preconditions_checkIndex modifies control flow even if the intrinsic bailed out

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 2c92d5ba10514d620c827fc06569dfa9cf1d0251
 * [Jatin Bhateja](https://openjdk.java.net/census#jbhateja) (@jatin-bhateja - Committer)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/114/head:pull/114` \
`$ git checkout pull/114`

Update a local copy of the PR: \
`$ git checkout pull/114` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/114/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 114`

View PR using the GUI difftool: \
`$ git pr show -t 114`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/114.diff">https://git.openjdk.java.net/jdk17/pull/114.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by sviswa7** [comment]
Created a separate RFE for test case https://bugs.openjdk.java.net/browse/JDK-8269107.

File: src/hotspot/share/opto/library_call.cpp [hunk 1]
```diff
@@ -1032,17 +1032,17 @@ bool LibraryCallKit::inline_preconditions_checkIndex(BasicType bt) {
                   Deoptimization::Action_make_not_entrant);
   }
 
+  if (stopped()) {
```
Comment (jatin-bhateja) [inline src/hotspot/share/opto/library_call.cpp:1035]: Hi @sviswa7, comments over both the short-circuiting conditions will help in better code readability.

Comment (sviswa7) [reply src/hotspot/share/opto/library_call.cpp:1035]: @jatin-bhateja Added comments, please take a look.

**Review by jatin-bhateja — APPROVED** [review]
Thanks @sviswa7

**Comment by sviswa7** [comment]
@vnkozlov Can I go ahead and integrate this into JDK 17?

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by vnkozlov** [comment]
Yes, you can and you did already 👍

**Pull request closed without merging**